Updatecheck fixes

Taloth opened the pull request:
@markus101 Haven't actually tested this integrated yet, but just wanted put it in a PR for some feedback.

Started out with some minor changes, but a check for writable folders were sprinkled all over the place, so put that in DiskProvider.
On linux I could probably check for permissions, but not on windows and not on unc and not on.... so I decided to just keep the 'write a file' check.

The last commit is about making the user aware the manual triggered install failed. We still don't report if the Update app failed, but at least we can check if the folder is writable and such.

I also intend to move the branch switch to InstallUpdate, makes more sense there.

markus101 commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -170,7 +182,12 @@ public void Execute(ApplicationUpdateCommand message)
+
+            if (!success)
+            {
+                throw new NzbDroneClientException(System.Net.HttpStatusCode.Conflict, "Failed to install update");
```
Not sure if conflict is the right status code here, but offhand I can't think of a better one.

Taloth replied:
It's a hack imo, we need a better way to report command failures to the UI.

markus101 replied:
Right now we do it through command progress messaging, I'm not sure that this exception will even make it out of the command executor.

markus101 commented on src/NzbDrone.Core/Update/UpdateCheckService.cs:
```diff
@@ -30,13 +31,13 @@ public class CheckUpdateService : ICheckUpdateService
         {
+            var latestAvailable = _updatePackageProvider.GetLatestUpdate(_configFileProvider.Branch, BuildInfo.Version);
+
             if (OsInfo.IsNotWindows && !_configFileProvider.UpdateAutomatically)
```
I was looking at this the other day and thought it was weird, this means people without auto update on don't get counted, doesn't it?

Taloth replied:
Exactly, so I swapped it around.

markus101 replied:
Okay good, I wasn't going crazy, which means we also might have a lot more users... on old versions.

kayone replied:
which we will never know about, since they won't get this update ;)

kayone replied:
We shall call them _The Lost Generation!_

markus101 replied:
maybe we will... if they manually update, one day.

kayone commented on src/NzbDrone.Common/Disk/DiskProviderBase.cs:
```diff
@@ -109,6 +109,23 @@ public bool FileExists(string path, StringComparison stringComparison)
+
+            try
+            {
+                var testPath = Path.Combine(path, "drone_test.txt");
```
`sonarr_write_test.txt`

kayone commented on src/NzbDrone.Common/Disk/DiskProviderBase.cs:
```diff
@@ -109,6 +109,23 @@ public bool FileExists(string path, StringComparison stringComparison)
+            try
+            {
+                var testPath = Path.Combine(path, "drone_test.txt");
+                File.WriteAllText(testPath, DateTime.Now.ToString());
```
I think we should make the text:
`this file was created to verify if {path} is writable. It should've been automatically deleted. Feel free to delete it.`

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -61,12 +62,20 @@ public class InstallUpdateService : IExecute<ApplicationUpdateCommand>, IExecute
+                {
+                    if (!_diskProvider.FolderWritable(_appFolderInfo.StartUpFolder))
+                    {
+                        throw new ApplicationException("Cannot install update because startup folder is not writable.");
```
include the path in the exception.

kayone commented on src/NzbDrone.Core/Update/UpdateCheckService.cs:
```diff
@@ -46,7 +47,7 @@ public UpdatePackage AvailableUpdate()
                     _logger.Info("Branch [{0}] is being redirected to [{1}]]", _configFileProvider.Branch, latestAvailable.Branch);
-                    var config = _configFileProvider.GetConfigDictionary();
+                    var config = new Dictionary<string, object>();
                     config["Branch"] = latestAvailable.Branch;
```
we should write a unit-test for `ConfigFileProvider` to make sure it will do what this excepts it to do.

Taloth commented:
Updated a few things.
Also moved the branch switch logic to InstallUpdate so the branch is only switched when a new version is installed instead only during the check.

Taloth commented on src/NzbDrone.Common.Test/ConfigFileProviderTest.cs:
```diff
@@ -141,8 +144,30 @@ public void SaveDictionary_should_save_proper_value()
+            int origSslPort = 20551;
+            int sslPort = 20552;
+
+            WithPartialDiskProvider();
```
A bit torn about this, I could also just setup the mock to return the correct Config xml code.

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -108,13 +119,35 @@ private void InstallUpdate(UpdatePackage updatePackage)
+            }
+        }
+
+        private void SwitchUpdatePackageBranch(UpdatePackage package)
```
The method name implies, switch is happening, we should either get rid of the if in this method, or call it `EnsureValidBranch` or something similar.

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -108,13 +119,35 @@ private void InstallUpdate(UpdatePackage updatePackage)
+                return;
             }
-            catch (Exception ex)
+            catch (UpdateFailedException ex)
```
What if it fails with another exception? the error message below seems pretty generic so I think we should catch all exceptions.

Taloth replied:
Previously a generic Exception would be caught by InstallUpdate and logged as Error, and then it swallowed.
Right now I let it get thrown up the stack and caught in the CommandExecutor, which logs it as Error too. So I didn't see any use in catching Exception specifically.

The only difference is that I catch UpdateFailedException and log it as Error, then it gets thrown again, and caught by the Execute(Command command) which turns it into a specific Failure (expected failure) and doesn't actually throw anymore. So all UpdateFailedExceptions get swallowed, but any other are allowed to get thrown all the way up.

I could leave out the entire try catch here and log in ExecuteInstallUpdate instead, if that is preferable.

kayone replied:
your second paragraph sounds like two component that shouldn't know about each other know why to much. this method is catching a specific exception because you know it get swallowed somewhere else.

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -108,13 +119,35 @@ private void InstallUpdate(UpdatePackage updatePackage)
+                }
+                catch (Exception e)
+                {
+                    _logger.ErrorException("Couldn't save the branch redirect.", e);
```
"Couldn't save branch redirect. {from}>{to}"

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -153,7 +186,29 @@ private void EnsureAppDataSafety()
                 _appFolderInfo.StartUpFolder.PathEquals(_appFolderInfo.AppDataFolder))
             {
-                throw new NotSupportedException("Update will cause AppData to be deleted, correct you configuration before proceeding");
+                throw new UpdateFailedException("Update will cause AppData to be deleted, correct you configuration before proceeding");
```
"You sonarr configuration is being stored in Application folder which will cause data lost during the upgrade. Please remove any symlinks or redirects before trying again."

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -164,13 +219,20 @@ public void Execute(ApplicationUpdateCommand message)
+
+            if (latestAvailable == null || latestAvailable.Hash != message.UpdatePackage.Hash)
+            {
+                throw new ApplicationException("Unknown or invalid update specified");
```
What does this mean? how can a user specify an update?

kayone replied:
also, null update and invalid hash warrant different error messages.

markus101 replied:
Previously we accepted a POST of whatever was submitted, which could have resulted in a malicious package being installed.

If the update isn't the latest the messaging should say that, entirely possible that it happens due to the user sitting on the update page for a while and a new update is available (develop in particular).

Taloth replied:
Mentioned this on Hangouts. We took the package sent back from the UI, including url and hash, which would allow anyone with access to the UI to upload any application they want and run it by submitting a package of their own.
This check simple verifies if the latest available update is the one the user clicked on and panics otherwise. It shouldn't happen in normal circumstances.

Taloth replied:
Should I split up the message? or just return on latestAvailable == null, or just always take latest and ignore the message.UpdatePackage entirely?

kayone replied:
We should update the code so it just uses the version sent from UI and ignores the hash and url.

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -153,7 +186,29 @@ private void EnsureAppDataSafety()
+            }
+            catch (UpdateVerificationFailedException ex)
+            {
+                message.Failed(ex, "Update verification failed");
```
needs a better error message.

Taloth replied:
It can't be too long coz it's shown on the UI notification. "Downloaded update package is corrupt"?

kayone replied:
Why can't it be too long, messenger doesn't seem to truncate it.

![image](https://cloud.githubusercontent.com/assets/412280/5906215/c0358d4c-a54a-11e4-8267-1396a7a4b9f8.png)

kayone commented:
pretty nice changes. 

I just feel like sometimes we try to be too brief in our error messages or tool-tips and we sacrifice them actually being useful. I would rather have a long error message that tells me why something happened and how I could fix it rather than a short one that only tells me something didn't work with no help on how to fix it.

markus101 commented on src/NzbDrone.Core/Messaging/Commands/Command.cs:
```diff
@@ -60,21 +60,21 @@ public void Start()
         }
 
-        public void Completed()
+        public void Completed(string message = "Completed")
```
:+1:

Taloth commented:
If we want longer explanations in the notification we should allow someone to click on it (or a button in the notifaction) which opens up a popup with a full explanation. But to make that useful we need some kind of knowledge-base api to supply additional details.

kayone commented:
why can't we just show the full message in the pop-up?

kayone commented on src/NzbDrone.Core/Update/InstallUpdateService.cs:
```diff
@@ -153,7 +176,32 @@ private void EnsureAppDataSafety()
             {
-                throw new NotSupportedException("Update will cause AppData to be deleted, correct you configuration before proceeding");
+                throw new UpdateFailedException("You Sonarr configuration ('{0}') is being stored in application folder ('{1}') which will cause data lost during the upgrade. Please remove any symlinks or redirects before trying again.", _appFolderInfo.AppDataFolder, _appFolderInfo.StartUpFolder);
+            }
```
we usually don't wrap values in logs with `()` we either use `[]` or `''`

kayone commented on src/NzbDrone.Core/Download/DownloadClientBase.cs:
```diff
@@ -102,22 +102,13 @@ protected ValidationFailure TestFolder(String folder, String propertyName, Boole
-                        DetailedDescription = "The folder you specified is not writable. Please verify the folder permissions for the user account that is used to execute NzbDrone."
-                    };
-                }
+                    DetailedDescription = "The folder you specified is not writable. Please verify the folder permissions for the user account that is used to execute NzbDrone."
```
"Sonarr"
also should include the user name like we do on other message.
